Implement 3D map viewers manager

NEDJIMAbelgacem opened the pull request:
## Description
Previously when a user closes a 3D map viewer then saves the project, the 3D map viewer will be lost.
This PR adds a 3D map views manager that takes care of removing 3D map views from the project and also renaming and duplication of 3D map views.

- 3D Map Views menu:

Previously, QGIS app view menu contained only "New 3D Map View" action. The PR replaces that action with "3D Map Views" sub menu, the sub menu contains the created map views for easy access, the "New 3D Map View" action and "Manage 3D Map Views". Screenshot of the "3D Map Views" menu :

![Screenshot from 2021-12-30 12-37-48](https://user-images.githubusercontent.com/29183781/147748673-950f8ad2-cfa5-440c-85fa-21e1c3aaff98.png)

- 3D Map Views Manager dialog:

The 3D Map Views Manager gives users the ability to open, duplicate, rename and remove 3D map views, Screenshot for the 3D map views manger:

![Screenshot from 2021-12-30 12-42-48](https://user-images.githubusercontent.com/29183781/147748935-3543debe-b637-49c9-9bfb-289f5e8df964.png)
The rename and duplicate buttons will ask the user for a new 3D Map viewer title.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

NEDJIMAbelgacem committed:
initial implementation

NEDJIMAbelgacem committed:
cleanup qgs3dmapviewsmanager

NEDJIMAbelgacem committed:
ask user for titles

NEDJIMAbelgacem committed:
fixes

NEDJIMAbelgacem committed:
fix removing closed 3d map view

NEDJIMAbelgacem committed:
add license

nirvn commented:
That's fantastic. Is there any shared code across 2d ans 3d map views to generalize this into a map views  manager for both types?

NEDJIMAbelgacem committed:
add ifdefs

NEDJIMAbelgacem commented:
> That's fantastic. Is there any shared code across 2d ans 3d map views to generalize this into a map views manager for both types?

Hi @nirvn 
The behaviour could be replicated for 2D map views with the exception that there is always a main map view that doesn't have a title in QGIS. I am not sure that it is a good idea to have a 2D views manager that has the ability to remove or rename map views since that might break some plugins in some way.

nirvn commented:
I've often found the destruction of 2d map views on closing the panel frustrating, just like their 3d counterparts. I can see why 3d views are extra frustrating due to a much larger amount of settings.

It'd still be nice IMHO. 

@nyalldawson , thoughts?

NEDJIMAbelgacem committed:
remove some unnecessary stuff

uclaros commented:
Most probably fixes #40753!

nyalldawson commented on src/app/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,52 @@
+    void removeClicked();
+    void renameClicked();
+  private:
+    QStringListModel mListModel;
```
This should instead be a pointer, with the QStringListModel parented to the dialog

nyalldawson commented on src/app/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,52 @@
+  private:
+    QStringListModel mListModel;
+
+    QMap<QString, QDomElement> *m3DMapViewsDom = nullptr;
+    QMap<QString, Qgs3DMapCanvasDockWidget *> *m3DMapViewsWidgets = nullptr;
```
```suggestion
    QMap<QString, QDomElement> m3DMapViewsDom;
    QMap<QString, Qgs3DMapCanvasDockWidget *> m3DMapViewsWidgets;
```

nyalldawson commented on src/app/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+  if ( m3DViewsListView->selectionModel()->selectedRows().isEmpty() )
+    return;
+
+  QString viewName = mListModel.stringList()[ m3DViewsListView->selectionModel()->selectedRows().at( 0 ).row() ];
```
You could take this directly from the data instead:

```suggestion
  QString viewName = m3DViewsListView->selectionModel()->selectedIndexes().at( 0 ).data( Qt::DisplayRole );
```

nyalldawson commented on src/app/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+  if ( m3DViewsListView->selectionModel()->selectedRows().isEmpty() )
+    return;
+
+  QString existingViewName = mListModel.stringList()[ m3DViewsListView->selectionModel()->selectedRows().at( 0 ).row() ];
```
```suggestion
  QString existingViewName = m3DViewsListView->selectionModel()->selectedIndexes().at( 0 ).data( Qt::DisplayRole );
```

nyalldawson commented on src/app/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+  if ( m3DViewsListView->selectionModel()->selectedRows().isEmpty() )
+    return;
+
+  QString viewName = mListModel.stringList()[ m3DViewsListView->selectionModel()->selectedRows().at( 0 ).row() ];
```
```suggestion
  QString viewName = m3DViewsListView->selectionModel()->selectedIndexes().at( 0 ).data( Qt::DisplayRole );
```

nyalldawson commented on src/app/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  QString viewName = mListModel.stringList()[ m3DViewsListView->selectionModel()->selectedRows().at( 0 ).row() ];
+  m3DMapViewsDom->remove( viewName );
+  Qgs3DMapCanvasDockWidget *w = m3DMapViewsWidgets->value( viewName, nullptr );
+  if ( w )
```
```suggestion
  if ( Qgs3DMapCanvasDockWidget *w = m3DMapViewsWidgets->value( viewName, nullptr ) )
```

nyalldawson commented on src/app/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,148 @@
+  if ( m3DViewsListView->selectionModel()->selectedRows().isEmpty() )
+    return;
+
+  QString oldTitle = mListModel.stringList()[ m3DViewsListView->selectionModel()->selectedRows().at( 0 ).row() ];
```
```suggestion
  QString oldTitle = m3DViewsListView->selectionModel()->selectedIndexes().at( 0 ).data( Qt::DisplayRole );
```

nyalldawson commented:
That was only a partial review, sorry. I think this needs reworking somewhat to remove the DOM element map as a the canonical place for storing the 3d map views, and instead there should be a small class QObject based class which tracks these. The dialog would ask the manager class to rename/duplicate/remove when needed, and the manager would have the logic to write the views to the project and read from them too.

(You could get away with copying a lot of this logic direct from QgsLayoutManager)

DelazJ commented:
Thanks @NEDJIMAbelgacem 
Very welcome. I recently discovered that map views (both 2d and 3d) were listed under the View-->Panels list, with buggy/unpredictable behavior (see #46446, see #46447, #46448). With this PR, are 3D map views removed from that list and does it fix any of the issues above?
I don't know how it could be done (given your concern with main map view handling), but a common manager for 2d and 3d views, as well as a dedicated list menu separating them from the other panels would indeed be a nice improvement, IMHO.

Will the view manager be closed only with the cross button or could you add a "close" button? That could (later) be seconded by a "help" button, same as with layout manager (which BTW lists both print layout and reports :wink:).

I know that makes that menu even longer but what about keeping the "new 3D map view" in the View menu and not as submenu?

This PR also fixes #46445, in another way than #46651. Do you think the fixes are not compatible?

NEDJIMAbelgacem committed:
Address Nyall reviews

NEDJIMAbelgacem committed:
fix 3D map view duplication

NEDJIMAbelgacem committed:
split 3dviewsmanager into 3dviews manager and dialog class

NEDJIMAbelgacem committed:
fix docs

NEDJIMAbelgacem committed:
fix ci build

NEDJIMAbelgacem committed:
fix sip

NEDJIMAbelgacem committed:
remove 3D views from view->panels and make checkable in view->3d views

NEDJIMAbelgacem committed:
use canvas names instead of open canvas names in new3DMapCanvas

NEDJIMAbelgacem commented:
> Very welcome. I recently discovered that map views (both 2d and 3d) were listed under the View-->Panels list, with buggy/unpredictable behavior (see #46446, see #46447, #46448). With this PR, are 3D map views removed from that list and does it fix any of the issues above?

I added some code to remove the 3D views from View-->Panels list and made the 3D views list in View-->3D Map Views checkable the way it was in View-->Panels. But it is a bit tricky to remove the 3D views from the pop up menu (right click on toolbar). As for 2D views, this PR is addressing the 3D views and if necessary 2D views manager will be implemented in a seperate PR.



> Will the view manager be closed only with the cross button or could you add a "close" button? That could (later) be seconded by a "help" button, same as with layout manager (which BTW lists both print layout and reports )

Sounds like a good idea 👍🏿 

> I know that makes that menu even longer but what about keeping the "new 3D map view" in the View menu and not as submenu?

I think keeping things contextual in the menu will be better. Let's see what other people think about it :) 


> This PR also fixes #46445, in another way than #46651. Do you think the fixes are not compatible?

I think from functional perspective is compatible, but it will require many changes to make the 2 PRs compatible.

NEDJIMAbelgacem commented:
> That was only a partial review, sorry. I think this needs reworking somewhat to remove the DOM element map as a the canonical place for storing the 3d map views, and instead there should be a small class QObject based class which tracks these. The dialog would ask the manager class to rename/duplicate/remove when needed, and the manager would have the logic to write the views to the project and read from them too.
> 
> (You could get away with copying a lot of this logic direct from QgsLayoutManager)

Thanks @nyalldawson !
I addressed your reviews and I reworked the code to move the DOM element map into a seperate manager class. Any other suggessions?

NEDJIMAbelgacem committed:
add close button

NEDJIMAbelgacem commented:
# Updated UI:
- Added checkboxes to indicate whether a 3D view is open and close or open directly from the menu.
- Added extra button to close the 3D view from the views manager dialog.

# Screenshots:

![Screenshot from 2022-01-02 22-09-52](https://user-images.githubusercontent.com/29183781/147889578-f3576bac-7df7-49aa-af62-85d3adce6420.png)
![Screenshot from 2022-01-02 22-08-29](https://user-images.githubusercontent.com/29183781/147889602-cd218353-98b4-408a-a51e-aa763c0732f6.png)

NEDJIMAbelgacem committed:
various fixes

NEDJIMAbelgacem committed:
fix tests

NEDJIMAbelgacem committed:
fix disabling buttons

NEDJIMAbelgacem committed:
fix crash

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+ *
+ * \brief Manages storage of a set of 3D views.
+ *
+ * QgsLayoutManager handles the storage, serializing and deserializing
```
s/QgsLayoutManager/.../

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+ *
+ * \since QGIS 3.24
+ */
+class CORE_EXPORT Qgs3DViewsManager : public QObject
```
Can we rename this as QgsMapViewManager, so that in future we can extend to include storage of 2d views as well? (also move to src/core/project accordingly)

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+    //! Removes and deletes all 3D views from the manager.
+    void clear();
+
+    //! Returns the DOM element representing the settings of the 3D view named \a name
+    QDomElement get3DViewSettings( const QString &name ) const;
```
```suggestion
    /**
     * Returns the DOM element representing the settings of the 3D view named \a name.
     *
     * \note Not available in Python bindings
    QDomElement get3DViewSettings( const QString &name ) const SIP_SKIP;
```

Let's skip this one in the Python bindings -- the raw DOM element doesn't have any real value there.

nyalldawson replied:
(The "get" prefix on the name here isn't QGIS standard, but I can see why it's required here. I'm fine with the exception here)

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+    //! Returns the DOM element representing the settings of the 3D view named \a name
+    QDomElement get3DViewSettings( const QString &name ) const;
+
+    //! Adds a new 3D view named \a name to the manager with the configuration DOM \a dom
+    void register3DViewSettings( const QString &name, const QDomElement &dom );
```
```suggestion
    /**
     * Adds a new 3D view named \a name to the manager with the configuration DOM \a dom
     *
     * \note Not available in Python bindings
     */
    void register3DViewSettings( const QString &name, const QDomElement &dom ) SIP_SKIP;
```

(for the same reason as above)

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+    QStringList get3DViewsNames() const;
+
+    //! Returns the list of configurations of 3D views added to the manager
+    QList<QDomElement> get3DViews() const;
```
SIP_SKIP this one

nyalldawson commented on src/core/3d/qgs3dviewsmanager.h:
```diff
@@ -0,0 +1,101 @@
+    void rename3DView( const QString &oldTitle, const QString &newTitle );
+
+    //! Sets the configuration of the 3D view named \a name to being opened
+    void viewOpened( const QString &name );
+
+    //! Sets the configuration of the 3D view named \a name to being closed
+    void viewClosed( const QString &name, const QDomElement &dom );
```
This API looks a bit odd. Could we go with something like `set3DViewInitiallyVisible( const QString& name, bool visible )` instead?

nyalldawson commented on src/core/3d/qgs3dviewsmanager.cpp:
```diff
@@ -0,0 +1,112 @@
+
+#include "qgs3dviewsmanager.h"
+#include "qgsproject.h"
+#include "qgsstyleentityvisitor.h"
```
```suggestion
```

nyalldawson commented on src/app/3d/qgs3dviewsmanagerdialog.cpp:
```diff
@@ -0,0 +1,200 @@
+  if ( widget )
+  {
+    widget->show();
+    QgsProject::instance()->setDirty();
```
```suggestion
```

Opening an existing view isn't really a change to the project -- I don't think it should cause a dirtying (no more so then opening/closing a layout designer window does)

NEDJIMAbelgacem replied:
The 3D views are automatically opened at the start of the project unlike layour designers and showing a hidden project changes whether it opens on start or not. I think it should be dirtying the project

nyalldawson replied:
I still disagree -- the purpose of this improvement is to separate the widget from the 3d map. So the widget state no longer becomes important anymore.

But happy for a second opinion here. Ping @wonder-sk @nirvn

nirvn replied:
I tend to agree with @nyalldawson here -- we don't dirty the project when we pan / zoom the map canvas, it makes sense to not dirty the project when opening / closing a 3D map view now that the panel visibility is dissociated from the 3D map view's existence. Similarly, we wouldn't dirty up a project when we toggle an attribute table.

That being said, I can see how someone can strongly feel in the opposite opinion :) as you're pointing out @NEDJIMAbelgacem , 3D map views aren't exactly project layouts / attribute tables since their visibility state are restored upon project load.

My vote would be a +0.25 towards no dirtying to keep the dirtying concept tied to _data/settings modification_ vs. state change.

I'm not being very helpful now am I? :wink:

NEDJIMAbelgacem replied:
One thing I want to add is that when the 3D view is closed, its settings are updated with the latest changes to the 3D Map view like the camera orientation and position. This is done so  that the next time the user opens the 3D map view will be back to the same configuration of when he closed it.

nirvn replied:
@NEDJIMAbelgacem , well then it seems that would likely dirty the project, so you can remove setDirty here :)

NEDJIMAbelgacem replied:
The is something else I want to say too: 
For now from the views manager standpoint the camera configuration and the rest of the configurations are not sperated and will need splitting the DOM into camera/non savable configuration and settings that are saved in the project.

@nirvn I'm a bit confused here, did you mean to keep or remove 😅 
I spoke with @wonder-sk and @uclaros and both of them seem to agree with my points - What do you think @nyalldawson ?

nyalldawson replied:
Well, a view change shouldn't result in project dirtying either (i.e. zooming into a 2d map or moving the 3d map camera). The new position should definitely be saved if the user clicks to save the project, but a view change alone shouldn't classify as a modification to the project. (Just the same way as MS word won't mark a document as dirty if you open and scroll through a few pages, zoom in/out, etc). 

That's the way 2d canvas currently works, so 3d should also follow this approach

nyalldawson commented on src/app/3d/qgs3dviewsmanagerdialog.cpp:
```diff
@@ -0,0 +1,200 @@
+    widget->close();
+  }
+
+  QgsProject::instance()->setDirty();
```
```suggestion
```

nyalldawson commented on src/app/3d/qgs3dviewsmanagerdialog.cpp:
```diff
@@ -0,0 +1,200 @@
+  }
+
+  QString viewName = current.data( Qt::DisplayRole ).toString();
+  bool isOpen = QgsProject::instance()->views3DManager()->get3DViewSettings( viewName ).attribute( QStringLiteral( "isOpen" ), QStringLiteral( "0" ) ).toInt() == 1;
```
this is messy -- we should keep the internal dom manipulation/attribute retrieval centralized in the view manager class only. (Otherwise this is highly likely to get out-of-sync with changes made in that class.)

Better though would be to keep QgisApp::instance() as responsible for tracking the open 3d dock widgets, and then here we could just call something like `QgisApp::instance()->is3DViewOpen( viewName )`. We could do this by adding a `QMap< QString, QPointer< Qgs3DMapCanvasDockWidget > > mOpen3DDocks` to QgisApp. This would also allow us to replace the `QgisApp::instance()->findChild<Qgs3DMapCanvasDockWidget *>( title + QStringLiteral( "DockObject" ) ) )` calls elsewhere here too, which also look a bit fragile.

NEDJIMAbelgacem replied:
Insetead of a Map I added a QSet to record open 3D views because by having a map we will need to add a rename function..

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -9961,6 +10063,53 @@ void QgisApp::populateLayoutsMenu( QMenu *menu )
+    // sort actions by text
+    std::sort( acts.begin(), acts.end(), cmpByText_ );
+  }
+  acts << mActionNew3DMapCanvas << mActionManage3DMapViews;
+  menu->addActions( acts );
```
```suggestion
  menu->addActions( acts );
  menu->addSeperator();
  menu->addAction( mActionNew3DMapCanvas );
  menu->addAction( mActionManage3DMapViews );
```

nyalldawson commented on src/ui/3d/qgs3dviewsmanagerdialog.ui:
```diff
@@ -0,0 +1,106 @@
+      </sizepolicy>
+     </property>
+     <property name="text">
+      <string>&amp;Show...</string>
```
```suggestion
      <string>&amp;Show</string>
```

nyalldawson commented on src/ui/3d/qgs3dviewsmanagerdialog.ui:
```diff
@@ -0,0 +1,106 @@
+      </sizepolicy>
+     </property>
+     <property name="text">
+      <string>&amp;Hide...</string>
```
```suggestion
      <string>&amp;Hide</string>
```

nyalldawson commented:
Thanks @NEDJIMAbelgacem , this is getting closer! :+1:

NEDJIMAbelgacem committed:
address Nyall reviews

NEDJIMAbelgacem committed:
Merge branch 'master' into map-3d-viewers-manager

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -755,6 +756,21 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
     QgsLayoutManager *layoutManager();
 
+    /**
+     * Returns the project's 3D views manager, which manages 3D views
```
```suggestion
     * Returns the project's map views manager, which manages map views (including 3d maps)
```

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -755,6 +756,21 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+    const QgsMapViewsManager *get3DViewsManager() const SIP_SKIP;
+
+    /**
+     * Returns the project's 3D views manager, which manages 3D views
```
```suggestion
     * Returns the project's map views manager, which manages map views (including 3d maps)
```

nyalldawson commented on src/core/project/qgsproject.h:
```diff
@@ -755,6 +756,21 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
+     * in the project.
+     * \since QGIS 3.24
+     */
+    QgsMapViewsManager *get3DViewsManager();
```
```suggestion
    QgsMapViewsManager *mapViewsManager();
```

NEDJIMAbelgacem committed:
rename stuff

NEDJIMAbelgacem committed:
rename stuff fix

NEDJIMAbelgacem committed:
remove dirtifying the project on hide & show and add proper project dirtyfing when the project changes

nyalldawson commented on src/3d/qgs3dmapsettings.cpp:
```diff
@@ -92,6 +108,7 @@ Qgs3DMapSettings::~Qgs3DMapSettings()
 
 void Qgs3DMapSettings::readXml( const QDomElement &elem, const QgsReadWriteContext &context )
 {
+  disconnectChangedSignalsToSettingsChanged();
```
Instead of the connect/disconnect dance here, you could use QgsProjectDirtyBlocker to instead prevent the project from being dirtied during the scope of this function.

NEDJIMAbelgacem committed:
use QgsProjectDirtyBlocker instead of connect disconnect dance

nyalldawson commented:
Lgtm!
